### PR TITLE
8264543: Cross modify fence optimization for x86

### PR DESCRIPTION
**Pull request by xbzhang99**
Intel introduced a new instruction “serialize” which ensures that all modifications to flags, registers, and memory by previous instructions are completed and all buffered writes are drained to memory before the next instruction is fetched and executed. It is a serializing instruction and can be used to implement cross modify fence (OrderAccess::cross_modify_fence_impl) more efficiently than using “cpuid” on supported 32-bit and 64-bit x86 platforms.

The availability of the SERIALIZE instruction is indicated by the presence of the CPUID feature flag SERIALIZE, bit 14 of the EDX register in sub-leaf CPUID:7H.0H.

https://software.intel.com/content/www/us/en/develop/download/intel-architecture-instruction-set-extensions-programming-reference.html

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8264543](https://bugs.openjdk.java.net/browse/JDK-8264543): Cross modify fence optimization for x86

### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to ef649f1d833d9d8f539428f19f98543e2ca90da8
 * [Thomas Schatzl](https://openjdk.java.net/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to ef649f1d833d9d8f539428f19f98543e2ca90da8
 * [Sandhya Viswanathan](https://openjdk.java.net/census#sviswanathan) (@sviswa7 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4848/head:pull/4848` \
`$ git checkout pull/4848`

Update a local copy of the PR: \
`$ git checkout pull/4848` \
`$ git pull https://git.openjdk.java.net/jdk pull/4848/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4848`

View PR using the GUI difftool: \
`$ git pr show -t 4848`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4848.diff">https://git.openjdk.java.net/jdk/pull/4848.diff</a>

</details>

### CODE REVIEW

**Comment by dholmes-ora** [comment]
Just to clarify for other reviewers, many of the small changes here are needed to allow orderAccess.hpp to be able to include vm_version.hpp. That required breaking a dependency on universe.hpp by moving some code to a .cpp file and adjusting the includes.

File: src/hotspot/os_cpu/windows_x86/orderAccess_windows_x86.hpp
Comment (dholmes-ora) [inline src/hotspot/os_cpu/windows_x86/orderAccess_windows_x86.hpp:31]: With Visual Studio this is already included by intrin.h

File: src/hotspot/os_cpu/windows_x86/orderAccess_windows_x86.hpp
Comment (dholmes-ora) [inline src/hotspot/os_cpu/windows_x86/orderAccess_windows_x86.hpp:64]: I cannot see this defined in the list of intrinsics available for Visual Studio:

https://docs.microsoft.com/en-us/cpp/intrinsics/x64-amd64-intrinsics-list?view=msvc-160

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
This seems fine for Linux-x86 but I have concerns about whether the Windows changes actually work with Visual Studio? And you have not added any support in orderAccess_bsd_x86.hpp.

Please enable GitHub Actions for your repo (or run the workflow manually) to ensure basic build and testing is performed.

Thanks,
David

File: src/hotspot/os_cpu/windows_x86/orderAccess_windows_x86.hpp
Comment (xbzhang99) [reply src/hotspot/os_cpu/windows_x86/orderAccess_windows_x86.hpp:64]: I was compiling using --with-toolchain-version=2019
_serialize() is defined here https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=seri&expand=6062

File: src/hotspot/os_cpu/windows_x86/orderAccess_windows_x86.hpp
Comment (xbzhang99) [reply src/hotspot/os_cpu/windows_x86/orderAccess_windows_x86.hpp:31]: ok. will remove it

File: src/hotspot/os_cpu/windows_x86/orderAccess_windows_x86.hpp
Comment (xbzhang99) [reply src/hotspot/os_cpu/windows_x86/orderAccess_windows_x86.hpp:63]: Also ran specjvm2008 successfully

Comment (dholmes-ora) [reply src/hotspot/os_cpu/windows_x86/orderAccess_windows_x86.hpp:63]: You are linking to Intel documentation of the Intel C++ compiler. This `_serialize()` intrinsic needs to be available in Microsoft Visual Studio.

**Review by dholmes-ora — CHANGES_REQUESTED** [review]

File: src/hotspot/os_cpu/windows_x86/orderAccess_windows_x86.hpp
Comment (dholmes-ora) [reply src/hotspot/os_cpu/windows_x86/orderAccess_windows_x86.hpp:63]: I ran it through our build system and it builds fine but I'd be happier if I could find some documentation that shows the availability of `_serialize()` in Visual Studio.

Comment (dholmes-ora) [reply src/hotspot/os_cpu/windows_x86/orderAccess_windows_x86.hpp:63]: Update: it builds fine with VS 2019, but not VS 2017. I don't know if we still expect to be able to build with VS 2017. If we do then the new code will need a compiler version guard on it.

Comment (dholmes-ora) [reply src/hotspot/os_cpu/windows_x86/orderAccess_windows_x86.hpp:63]: Update: VS 2017 is the minimum supported compiler version so the new code will need to be enabled only for VS 2019.

File: src/hotspot/os_cpu/windows_x86/orderAccess_windows_x86.hpp
Comment (dholmes-ora) [inline src/hotspot/os_cpu/windows_x86/orderAccess_windows_x86.hpp:64]: 16.7.2 is the earliest I could look at, but it could be supported earlier. Also _MSC_VER of 1927 would start with 16.7.0 so it wouldn't quite be right to test for that if it only come in with 16.7.2.

Typo: ntrinsic

File: src/hotspot/os_cpu/windows_x86/orderAccess_windows_x86.hpp
Comment (dholmes-ora) [inline src/hotspot/os_cpu/windows_x86/orderAccess_windows_x86.hpp:66]: _xx ???

File: src/hotspot/os_cpu/windows_x86/orderAccess_windows_x86.hpp
Comment (xbzhang99) [reply src/hotspot/os_cpu/windows_x86/orderAccess_windows_x86.hpp:64]: I changed to 1928 to be on the safe side

**Comment by dholmes-ora** [comment]
Thanks Xubo, I will run this through our test system using VS 2019 and VS 2017 before approving. You will still need a second reviewer after that.

David

**Comment by dholmes-ora** [comment]
@xbzhang99 what testing have you done on this?

All the JVMCI tests are failing as JVMCI needs to know about the new CPU feature bit for some reason:
`JVMCIError: Missing CPU feature constants: [SERIALIZE]`
I'm trying to find out what needs to be done to fix this.

**Comment by dholmes-ora** [comment]
You need this additional patch:
diff --git a/src/jdk.internal.vm.ci/share/classes/jdk.vm.ci.amd64/src/jdk/vm/ci/amd64/AMD64.java b/src/jdk.internal.vm.ci/share/classes/jdk.vm.ci.amd64/src/jdk/vm/ci/amd64/AMD64.java
index 3fc5b6e7936..f22e120ae53 100644
--- a/src/jdk.internal.vm.ci/share/classes/jdk.vm.ci.amd64/src/jdk/vm/ci/amd64/AMD64.java
+++ b/src/jdk.internal.vm.ci/share/classes/jdk.vm.ci.amd64/src/jdk/vm/ci/amd64/AMD64.java
@@ -220,6 +220,7 @@ public class AMD64 extends Architecture {
         AVX512_VBMI2,
         AVX512_VBMI,
         HV,
+        SERIALIZE,
     }
 
     private final EnumSet<CPUFeature> features;

**Comment by dholmes-ora** [comment]
Our testing passed with the additional patch.

**Review by dholmes-ora — APPROVED** [review]
These changes seem fine to me now.

A second review is still needed.

Thanks,
David

**Review by tschatzl — APPROVED** [review]
Lgtm apart from the "ntrinsic" typo mentioned by @dholmes

**Review by sviswa7 — APPROVED** [review]

**Comment by sviswa7** [comment]
The patch looks good to me.

**Comment by xbzhang99** [comment]
Hi @dholmes-ora
could you sponsor this PR?
Thanks!

**Pull request closed without merging**
